haystack_experimental/core/pipeline/breakpoint.py

```diff
@@ -0,0 +1,317 @@
+
+    # pipeline_state has the necessary keys
+    pipeline_state = state["pipeline_state"]
+    required_pipeline_keys = {"inputs", "component_visits", "ordered_component_names"}
+    missing_pipeline = required_pipeline_keys - pipeline_state.keys()
+    if missing_pipeline:
+        raise ValueError(f"Invalid pipeline_state: missing required keys {missing_pipeline}")
+
+    # component_visits and ordered_component_names must be consistent
+    components_in_state = set(pipeline_state["component_visits"].keys())
+    components_in_order = set(pipeline_state["ordered_component_names"])
+
+    if components_in_state != components_in_order:
+        raise ValueError(
+            f"Inconsistent state: components in pipeline_state['component_visits'] {components_in_state} "
+            f"do not match components in ordered_component_names {components_in_order}"
+        )
+
+    logger.info("Passed resume state validated successfully.")
+
+
+def load_state(file_path: Union[str, Path]) -> Dict[str, Any]:
+    """
+    Load a saved pipeline state.
+
+    :param file_path: Path to the state file
+    :returns:
+        Dict containing the loaded state
+    """
+
+    file_path = Path(file_path)
+
+    try:
+        with open(file_path, "r", encoding="utf-8") as f:
+            state = json.load(f)
+    except FileNotFoundError:
+        raise FileNotFoundError(f"File not found: {file_path}")
+    except json.JSONDecodeError as e:
+        raise json.JSONDecodeError(f"Invalid JSON file {file_path}: {str(e)}", e.doc, e.pos)
+    except IOError as e:
+        raise IOError(f"Error reading {file_path}: {str(e)}")
+
+    try:
+        _validate_resume_state(state=state)
+    except ValueError as e:
+        raise ValueError(f"Invalid pipeline state from {file_path}: {str(e)}")
+
+    logger.info(f"Successfully loaded pipeline state from: {file_path}")
+    return state
+
+
+def _save_state(
+    inputs: Dict[str, Any],
+    component_name: str,
+    component_visits: Dict[str, int],
+    callback_fun: Optional[Callable[..., Any]] = None,
+    debug_path: Optional[Union[str, Path]] = None,
+    original_input_data: Optional[Dict[str, Any]] = None,
+    ordered_component_names: Optional[List[str]] = None,
+) -> Dict[str, Any]:
+    """
+    Save the pipeline state to a file.
+
+    :param inputs: The current pipeline state inputs.
+    :param component_name: The name of the component that triggered the breakpoint.
+    :param component_visits: The visit count of the component that triggered the breakpoint.
+    :param callback_fun: A function to call with the saved state.
+    :param debug_path: The path to save the state to.
+    :param original_input_data: The original input data.
+    :param ordered_component_names: The ordered component names.
+    :raises:
+        Exception: If the debug_path is not a string or a Path object, or if saving the JSON state fails.
+
+    :returns:
+        The saved state dictionary
```

Comment:
Could we also list all the keys and their respective values in this docstring like we do in the `:returns:` for our components?